tests/unit/cluster/atomic-slot-migration.tcl

```diff
@@ -99,29 +99,54 @@ proc populate_slot {num args} {
     R $idx deferred 0
 }
 
+# Return 1 if all instances are idle
+proc asm_all_instances_idle {total} {
+    for {set i 0} {$i < $total} {incr i} {
+        if {[CI $i slot_migration_active_tasks] != 0} { return 0 }
+        if {[CI $i slot_migration_active_trim_running] != 0} { return 0 }
+    }
+    return 1
+}
+
 # Wait for all ASM tasks to complete in the cluster
 proc wait_for_asm_done {} {
     set total_instances [expr {$::cluster_master_nodes + $::cluster_replica_nodes}]
 
-    for {set i 0} {$i < $total_instances} {incr i} {
-        wait_for_condition 1000 10 {
-            [CI $i slot_migration_active_tasks] == 0 &&
-            [CI $i slot_migration_active_trim_running] == 0
-        } else {
+    wait_for_condition 1000 10 {
+        [asm_all_instances_idle $total_instances] == 1
+    } else {
+        # Print the number of active tasks on each instance
+        for {set i 0} {$i < $total_instances} {incr i} {
             set migration_count [CI $i slot_migration_active_tasks]
             set trim_count [CI $i slot_migration_active_trim_running]
-            fail "ASM tasks did not complete on instance $i: migration_tasks=$migration_count, trim_tasks=$trim_count"
+            puts "Instance $i: migration_tasks=$migration_count, trim_tasks=$trim_count"
         }
+        fail "ASM tasks did not complete on all instances"
```

Comment:
There was a race condition in this proc. If one instance holds true (no task on it), we move on and just wait on the other instance that has a task. If we see any instance has a task, we should loop over all the instances from the beginning. Because one instance may start the work later than the other, previously we couldn't catch this.